WebApp-GroupClaims-DotNet/Utils/AuthenticationHelper.cs

```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using System.Web;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using ClaimTypes = System.IdentityModel.Claims.ClaimTypes;
+
+namespace WebApp_GroupClaims_DotNet.Utils
+{
+    public class AuthenticationHelper
+    {
+        public const string OAuth2GrantTypeJwtBearer = "urn:ietf:params:oauth:grant-type:jwt-bearer";
+
+        public TokenCache TokenCache { get; set; }
+
+        public string Authority { get; set; }
+
+        public AuthenticationHelper(string authority, TokenCache tokenCache)
+        {
+            this.Authority = authority;
+            this.TokenCache = tokenCache;
+        }
+
+        /// <summary>
+        /// Get Token for User.
+        /// </summary>
+        /// <returns>Token for user.</returns>
+        public async Task<string> GetAccessTokenForUserAsync(string resourceId, string replyUrl, string userIdentifier = "")
+        {
+            AuthenticationContext authContext = new AuthenticationContext(this.Authority, this.TokenCache);
+            AuthenticationResult authResult = null;
+            ClientCredential credential = new ClientCredential(AppConfig.ClientId, AppConfig.AppKey);
+            try
+            {
+                authResult = await authContext.AcquireTokenSilentAsync(resourceId, credential, new UserIdentifier(userIdentifier, UserIdentifierType.UniqueId));
+            }
+            catch (AdalSilentTokenAcquisitionException)
+            {
+                //authResult = await authContext.AcquireTokenAsync(resourceId, AppConfig.ClientId, new Uri(replyUrl), new PlatformParameters(PromptBehavior.Auto), new UserIdentifier(userIdentifier, UserIdentifierType.OptionalDisplayableId));
+                authResult = await authContext.AcquireTokenAsync(resourceId, AppConfig.ClientId, new Uri(AppConfig.PostLogoutRedirectUri), new PlatformParameters(PromptBehavior.Auto));
```

Comment:
This override (which is an interactive / For public client apps) cannot be used in a Web App!
It's going (worse case) to attempt to popup a Window Forms popup on the web server :)

Here no choice but go with OBO